contracts/javascore/ibc/src/main/java/ibc/ics20/app/ICS20Bank.java

```diff
@@ -0,0 +1,104 @@
+        Context.require(account != ICS20Transfer.ZERO_ADDRESS, "ICS20Bank: balance query for the zero address");
+
+        // Assuming the denomination is a valid key in the balances mapping
+        return balances.at(denom).getOrDefault(account, BigInteger.ZERO);
+    }
+
+    @External
+    public void transferFrom(Address from, Address to, String denom, BigInteger amount) {
+        Context.require(to != ICS20Transfer.ZERO_ADDRESS, "ICS20Bank: balance query for the zero address");
+        Address caller = Context.getCaller();
+        Context.require(from.equals(caller) || hasRole(OPERATOR_ROLE_ID, caller), "ICS20Bank: caller is not owner nor approved");
+        BigInteger fromBalance = balanceOf(from, denom);
+        Context.require(amount.compareTo(BigInteger.ZERO) > 0, "ICS20Bank: transfer amount must be greater than zero");
+        Context.require(fromBalance.compareTo(amount) >= 0, "ICS20Bank: insufficient balance for transfer");
+
+        balances.at(denom).set(from, fromBalance.subtract(amount));
+        balances.at(denom).set(to, balanceOf(to, denom).add(amount));
+    }
+
+    @External
+    public void mint(Address account, String denom, BigInteger amount) {
+        Context.require(hasRole(OPERATOR_ROLE_ID, Context.getCaller()), "ICS20Bank: must have minter role to mint");
+        Context.require(account != ICS20Transfer.ZERO_ADDRESS, "ICS20Bank: mint to the zero address");
+        Context.require(amount.compareTo(BigInteger.ZERO) > 0, "ICS20Bank: mint amount must be greater than zero");
+        _mint(account, denom, amount);
+    }
+
+    @External
+    public void burn(Address account, String denom, BigInteger amount) {
+        Context.require(hasRole(OPERATOR_ROLE_ID, Context.getCaller()), "ICS20Bank: must have minter role to mint");
+        Context.require(amount.compareTo(BigInteger.ZERO) > 0, "ICS20Bank: mint amount must be greater than zero");
+        _burn(account, denom, amount);
+    }
+
+    @External
+    public void deposit(Address tokenContract, BigInteger amount, Address receiver){
+        Context.require(tokenContract.isContract(), "ICS20Bank: tokenContract is not a contract");
+        Context.call(tokenContract, "transferFrom", Context.getCaller(), Context.getAddress(), amount);
+        _mint(receiver, tokenContract.toString(), amount);
+    }
+
+    @External
+    public void withdraw(Address tokenContract, BigInteger amount, Address receiver){
+        Context.require(tokenContract.isContract(), "ICS20Bank: tokenContract is not a contract");
+        _burn(receiver, tokenContract.toString(), amount);
+        Context.call(tokenContract, "transfer", receiver, amount);
+    }
+
+    private void _mint(Address account, String denom,  BigInteger amount){
+        balances.at(denom).set(account, balanceOf(account, denom).add(amount));
+    }
+
+    private void _burn(Address account, String denom,  BigInteger amount){
+        BigInteger accountBalance = balanceOf(account, denom);
```

Comment:
parameter validation is done in inner function in _burn but not in _mint. We should stick with one strategy so that it will be less likely to miss validation or have double validation.

Reply:
need to check balance before burning only so there is the check only on _burn than mint